pkg/operator/constants/constants.go

```diff
@@ -118,6 +122,12 @@ const (
 	// KubevirtCloudCredSecretName is the name of the secret where credentials
 	// for Kubevirt are stored.
 	KubevirtCloudCredSecretName = "kubevirt-credentials"
+
+	// UpgradeableNextRelease is the version CCO will calculate whether it can upgrade to.
+	UpgradeableNextRelease = "4.7"
```

Comment:
I think we could make it current +1. Need to think about how this would work when we need to merge a commit into 4.7 (which should be benign in 4.7) before it gets into 4.6...

The annotation is only required when the upgradeParams.MissingAnnotationMessage is a non empty string. The string is the message to communicate what is expected from the user before setting the annotation to get CCO to Upgradeable=True.

Yes, for EUS, you would need to resolve any change from one release to the next. Without that, we would have to encode in .5 all the knowledge about getting from .5 to .8 (and a user might only be interested in the info for .5 to .6. This might be doable if .5.x is the last non-EUS update, and .5.y (and everything that follows) is exclusively EUS updates. Then it could make sense to encode all that information to upgrade to .8 in .5.y.

We could absolutely do the semver calculation. The challenge is communicating all of the info that might be needed to get from currentEUS to nextEUS. Perhaps the message is `Go read https://redhat.com/upgrading-4.5-to-4.8.html` ?

Reply:
See if this makes sense. If we make it current+.1

Cluster just upgraded to 4.8 by setting the annotation. To be more concrete, the user was instructed to add the s3:ListMultiPartUploadParts permission, and then the user set the annotation to unblock the upgrade to 4.8.

Upgradeable() in 4.8 (which 4.7 simply backported the check from (assuming we didn't know so far in advance that this new permission was needed until 4.8 was near release)) now sees that annotation indicating 4.8 (which of course is < 4.8 + .1). Now after a successful upgrade to 4.8, CCO marks things Upgradeable=False indicating the same steps (that were supposed to be for 4.8) are needed before upgrading to 4.9.

This scenario depends on the assumption that code needs to land in 4.8 before we can backport the 4.8 check into 4.7. If we're allowed to directly commit future-looking checks into 4.7 (bypassing 4.8) then we can avoid this.

Reply:
With the data defined in the cred request itself, and documented steps for how to upgrade, adding a third place where we convey the specifics is getting to be a lot to keep in sync. I'd be tempted to drop the idea of communicating exactly what needs to be done in the upgradeable message, it's fine for one permission, but it could be 20 across a bunch of cred requests someday. I think your idea of linking to docs that show how to extract the info from a release image *generically* would be ideal. The user should get familiar with that process if they're in manual mode. And of course ideally I'd love to let them have a tool to do it.

WDYT if it just said:  "You have not yet manually prepared credentials/permissions for release x.y. Please see https://redhat.com/manual-credentials-upgrade.html." And there we could show how to run the tool, or do it manually until it exists.

Tool would also have the same problem of not knowing what user identity belongs to a cred request, though we might be able to figure it out, and maybe even populate it.

Reply:
I do like this. It forces us to make sure the docs are in good shape (if they aren't already). All we have to do to be ready for a release is know whether there were any changes in the upcoming release, and we can turn on the "please annotate things before upgrading" path if something did change.

For STS, the tool (assuming the tool also made the IAM Roles) will know the names of the IAM Roles as they are named using a repeatable scheme (clusterInfra-targetNamespace-targetName https://github.com/sjenning/sts-preflight/blob/master/pkg/iamroles/iamrole.go#L68 ). I don't believe we have any plans to create tooling for the non-STS case so that situation would really be a "make sure you read the docs" situation.

Reply:
Should we defer the assumed target release or get it into this PR?

Reply:
This is the item I keep coming back to. How can we write this in a way that is backportable so that it works in 4.x (ie. set the Upgradeable=False until annotation is seen) and 4.x+1 (ie. don't set Upgradeable=False b/c the fact that we are in 4.x+1 means we must have already made the annotation to get to 4.x+1) w/o hardcoding these values?

Reply:
I'm thinking:

- current := 4.6 (introspect to find out)
- next := current+1 major (i.e. 4.7)
- if CR has no annotation, upgradable = false
- if CR has annotation that is semver >= next, upgradeable=true (i.e. annotation of 4.7, 4.8, 4.9, would all mark upgradable true if you're on 4.6)
- if CR has annotation that is semver < next, upgradeable=false (i.e. annotation of 4.6 would mark false if you're on 4.6)

Am I missing a case in there?

Reply:
I see what you mean now. 

We could take the stance, you need to get in the habit of running this tool to reconcile your creds, regardless if the release needs anything or not. This makes the code a lot simpler and gets users into good habits. Being upgradeable false in manual mode immediately though is something I'd consider to be fairly normal.

Or, we could encode info as to which specific releases do or don't have a problem on a platform. The AWS actuator could know a list of releases where changes are needed, i.e. ["4.7, "4.9"]. That's pretty much back to hardcoded which I see now is what you're thinking.

Reply:
I suppose the decision is whether Upgradeable=False should be the normal state of a cluster in Manual mode. Do we bring this up in our arch meeting?

Reply:
Yup lets get it onto the agenda for next week.